qemu/qemu.go

```diff
@@ -646,6 +647,400 @@ const (
+	Rerror DriveRWErrorsAction
+}
+
+// DriveRWErrorsAction is action to take on rw errors
+type DriveRWErrorsAction string
+
+const (
+	// DriveRWErrorsIgnore ignores the error
+	DriveRWErrorsIgnore DriveRWErrorsAction = "ignore"
+
+	// DriveRWErrorsStop stops Qemu
+	DriveRWErrorsStop DriveRWErrorsAction = "stop"
+
+	// DriveRWErrorsReport reports the error
+	DriveRWErrorsReport DriveRWErrorsAction = "report"
+
+	// DriveRWErrorsEnospc pauses Qemu only if the host disk is full
+	DriveRWErrorsEnospc DriveRWErrorsAction = "enospc"
+)
+
+// Valid returns true if the drive rw error is valid, else false
+func (d *DriveRWErrors) Valid() bool {
+	switch d.Werror {
+	case DriveRWErrorsIgnore, DriveRWErrorsStop, DriveRWErrorsReport, DriveRWErrorsEnospc:
+		break
+	default:
+		return false
+	}
+
+	switch d.Rerror {
+	case DriveRWErrorsIgnore, DriveRWErrorsStop, DriveRWErrorsReport, DriveRWErrorsEnospc:
+		break
+	default:
+		return false
+	}
+
+	return true
+}
+
+// Drive represents a qemu drive.
+type Drive struct {
+	// File defines which disk image to use with this drive
+	File string
+
+	// If defines on which type on interface the drive is connected
+	// Available types are: ide, scsi, sd, mtd, floppy, pflash, virtio
+	If DriveInterface
+
+	// Connection defines where is connected the drive
+	Connection *DriveConnection
+
+	// Index where is connected the drive is connected
+	Index string
+
+	// Media defines the type of the media
+	// Available types are: disk, cdrom
+	Media DriveMedia
+
+	// Geometry forces hard disk 0 physical geometry
+	Geometry *HardDiskGeometry
+
+	// Snapshot controls snapshot mode for the given drive
+	// Available values are: on, off
+	Snapshot DriveState
+
+	// Cache controls how the host cache is used to access block data
+	// Available types are: none, writeback, unsafe, directsync, writethrough
+	Cache DriveCacheMode
+
+	// AIO selects between pthread based disk I/O and native Linux AIO
+	// Available types are: threads, native
+	AIO DriveAIO
+
+	// Discard controls whether discard requests are ignored or passed to the filesystem
+	// Available values are: ignore, off, unmap, on
+	Discard DriveState
+
+	// Format specifies which disk format will be used rather than detecting the format
+	Format string
+
+	// Serial specifies the serial number to assign to the device
+	Serial string
+
+	// Addr specifies the controller's PCI address (if=virtio only)
+	Addr string
+
+	// Errors specifies which action to take on write and read errors
+	Errors *DriveRWErrors
+
+	// Readonly opens drive file as read-only
+	// Available values are: on, off
+	Readonly DriveState
+
+	// CopyOnRead enables whether to copy read backing file sectors into the image file
+	// Available values are: on, off
+	CopyOnRead DriveState
+
+	// DetectZeroes enables the automatic conversion of plain zero writes by the OS to driver specific optimized zero write
+	// Available values are: off, on, unmap
+	DetectZeroes DriveState
+}
```

Comment:
Overall I find this structure to be a little too complex. In my opinion one of the nice govmm feature is that it simplifies the QEMU interface by trying to only expose the most useful and used options. I think this structure could be reduced by a fair amount, first of all to meet our immediate needs. We can *expand* it later if there's some real need for things like `DriveGeometry` and other options. So my advice is to make this structure fits your needs first, while keeping in mind that it could be expanded further down the road.

Reply:
I included all possible option to play with them later